RecoTauTag/RecoTau/plugins/DeepTauId.cc

```diff
@@ -415,6 +421,20 @@ namespace {
     }
   }  // namespace dnn_inputs_2017_v2
 
+  float getTauID(const pat::Tau& tau, const std::string& tauID, float default_value = -999.) {
+    static tbb::concurrent_unordered_set<std::string> isFirstWarning;
+    if (tau.isTauIDAvailable(tauID)) {
+      return tau.tauID(tauID);
+    } else {
+      if (isFirstWarning.insert(tauID).second) {
+        edm::LogWarning("DeepTauID") << "Warning in <getTauID>: No tauID '" << tauID
```

Comment:
Isn't it worth a warning if an input ID is requested but not available? If it wasn't for development but for final usage it should even throw an exception (but of course be properly configured such that it doesn't happen). I'm not sure, in which exact context this became necessary, @veelken might want to add sth.

Reply:
OK, it sounds like it's definitely fine for a non-zero debug_level.
I understood that this was a proposal for a production setup as well, in which case we'll just be making some repeated noise in every job.
How frequent are these now for existing workflows?

Reply:
Ah ok, I understand. These developments are preparations for DeepTau at HLT and since this functionality concerns PAT taus I'm pretty sure that this is just to facilitate supporting studies. I don't think that there is a workflow in production that should trigger these warnings. @veelken could you say more about the exact use cases, please?

Reply:
I confirm that these warnings should not trigger during normal production. When developing the Phase-2 HLT tau trigger, I ran the DeepTau algorithm on PAT taus and ran into this error. I prefer that we print an error message in this case.

Reply:
I see, so if in the production case we expect to require all inputs/IDs present, perhaps it's better to have an exception for the default behavior and print a warning or error as you already have for the debugging/testing purposes.

Reply:
Is there actually a use case where the output is usable even if this warning was triggered @veelken ? Maybe we could turn it into throwing an exception for all cases where an ID is not available?

Reply:
Hi, in my opinion it will be fine to throw and exception in case the DeepTauId code is run in "production" mode and print the warning (without throwing an exception) only in case the DeepTauId code is run in development/"debug" mode.

The reason why I added the warning is that the DeepTauId code requires many tau ID discriminators as input, which didn't exist in the HLT tau trigger object when I started developing the Phase-2 HLT tau trigger. I later added the full set of tau ID discriminators to the Phase-2 HLT tau trigger, so that the Phase-2 HLT tau trigger can now be run in "production" mode (i.e. with the exception throws in case a tau ID discriminator does not exist).

Reply:
@slava77 how would we switch between the two cases? Is there a suitable flag existing?

Reply:
here we are talking about a standalone function; so I guess this means
```C++
   float getTauID(const pat::Tau& tau, const std::string& tauID, float default_value = -999., bool doThrow = true) {
```

Reply:
I guess my question was not clear. I mean whether there is already some flag in cmssw that you had in mind and that can be used to distinguish between let's say official production and private runs (whereas the latter is not necessarily the same as development). Since if we need to add a new option to the deepTauID config to steer this switch, we will end up in the situation that a developer runs the producer which will crash by default if an input ID is missing. Then the developer looks into the code and sees that he could set the option. But with about the same effort he could also turn the exception into cout for his development. Therefore it's maybe not worth to blow up the code with such a switch.